### PR TITLE
Create Bihemispheric Normorphic Network

### PR DESCRIPTION
**Pull request by NaderMalki**
#Using this network without permission #from its owner and inventor, Nader Maleki, will result in WIPO rules.#
PCT/IR2025/050026.nadar.malki#
Bihemispheric Normorphic Network#
import numpy as np
import numpy as np
import tensorflow as tf
from tensorflow.keras.layers import Layer, Dense, Input
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.datasets import load_iris
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# ----------------------------
# 1. Data
# ----------------------------
iris = load_iris()
X, y = iris.data, iris.target
X = StandardScaler().fit_transform(X)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
y_train = tf.keras.utils.to_categorical(y_train, 3)
y_test = tf.keras.utils.to_categorical(y_test, 3)

# ----------------------------
# 2. Advanced input layer: attention + modulation + two projections
# ----------------------------
class InputProcessor(Layer):
def init(self, proj_dim=8, kwargs):
super(InputProcessor, self).__init__(kwargs)
self.proj_dim = proj_dim
# Two projection heads
self.proj1 = Dense(proj_dim, activation='linear', name='proj_linear')
self.proj2 = Dense(proj_dim, activation='relu', name='proj_nonlinear')
# Attention weights for modulation
self.att_weight = self.add_weight(shape=(proj_dim,), initializer='ones', trainable=True)

def call(self, x):
# Filter/normalization already done
p1 = self.proj1(x) # Weight vector 1
p2 = self.proj2(x) # Weight vector 2
# Modulation with dynamic attention
modulated = p1 * tf.nn.sigmoid(self.att_weight) + p2 * (1 - tf.nn.sigmoid(self.att_weight))
return p1, p2, modulated

# ----------------------------
# 3. Hyperbolic Block (simulated)
# ----------------------------
def hyperbolic_activation(x):
"""Approximate the activator in hyperbolic space (Poincaré Ball)"""
norm = tf.norm(x, axis=-1, keepdims=True)
eps = 1e-6
return tf.math.tanh(norm + eps) * (x / (norm + eps))

class HyperbolicBlock(Layer):
def init(self, units=8, kwargs):
super(HyperbolicBlock, self).__init__(kwargs)
self.dense = Dense(units, activation='linear')

def call(self, x):
x = self.dense(x)
return hyperbolic_activation(x)

# ----------------------------
# 4. Dynamic attention mechanism between blocks
# ----------------------------
class CrossAttentionFusion(Layer):
def init(self, kwargs):
super(CrossAttentionFusion, self).__init__(kwargs)
self.w_q = Dense(8) # query
self.w_k = Dense(8) # key
self.w_v = Dense(8) # value

def call(self, linear_out, hyper_out):
# query from linear block, key/value from hyperbolic block
Q = self.w_q(linear_out)
K = self.w_k(hyper_out)
V = self.w_v(hyper_out)

# simple attention
attn_scores = tf.matmul(Q, K, transpose_b=True) / tf.math.sqrt(8.0)
attn_weights = tf.nn.softmax(attn_scores, axis=-1)
attended = tf.matmul(attn_weights, V)

# Final fusion
fused = linear_out + attended
return fused

# ----------------------------
# 5. Model assembly
# ----------------------------
inputs = Input(shape=(4,), name='input')

# Step 1: Input processing
proj1, proj2, modulated = InputProcessor(proj_dim=8, name='input_processor')(inputs)

# Step 2: Two parallel blocks
linear_branch = Dense(8, activation='linear', name='linear_branch')(modulated)
hyper_branch = HyperbolicBlock(units=8, name='hyperbolic_branch')(modulated)

# Step 3: Dynamic attention information exchange
fused = CrossAttentionFusion(name='cross_attention')([linear_branch, hyper_branch])

# Step 4: Output
outputs = Dense(3, activation='softmax', name='output')(fused)

model = Model(inputs=inputs, outputs=outputs)

# Compile
model.compile(
optimizer=Adam(learning_rate=0.001),
loss='categorical_crossentropy',
metrics=['accuracy']

print("✅ Model architecture built successfully!")
model.summary()

# ----------------------------
# 6. Training and testing
# ----------------------------
history = model.fit(
X_train, y_train,
validation_data=(X_test, y_test),
epochs=80,
batch_size=8,
verbose=1
)
import numpy as np
import tensorflow as tf
from sklearn.metrics import accuracy_score, precision_recall_fscore_support
import time
import matplotlib.pyplot as plt

# ----------------------------
# 1. Simulate piezoelectric data (same as your code)
# ----------------------------
time_sim = np.linspace(0, 2, 1000) # 2 seconds for better testing
sigma = 1e6 * np.abs(np.sin(2 * np.pi * 5 * time_sim))
d33 = 2.5e-10
t = 0.001
V = d33 * sigma * t # Raw voltage

# Create label: 1 if sigma > 0.8e6 (bounce), 0 otherwise
y_true = (sigma > 0.8e6).astype(int)

# Add realistic noise (for testing stability)
np.random.seed(42)
V_noisy = V + np.random.normal(0, np.std(V)*0.1, V.shape)

# ----------------------------
# 2. Preprocessing: Edge Detection with Gradient
# ----------------------------
def detect_edges(signal, threshold=1e-8):
dv_dt = np.gradient(signal, time_sim)
return (np.abs(dv_dt) > threshold).astype(int)

# Optimized threshold for our data
edge_pred = detect_edges(V_noisy, threshold=5e-9)

# ----------------------------
# 3. Evaluation Criteria
# ----------------------------
# a) Accuracy
acc = accuracy_score(y_true, edge_pred)
precision, recall, f1, _ = precision_recall_fscore_support(y_true, edge_pred, average='binary')

# b) Speed ​​(detection latency)
# Assumption: system samples every 10ms → latency = 10ms
latency_ms = 10 # milliseconds

# C) Training (simple here, but if it was a SNN with learning, we would measure the convergence rate)
trainable = "yes (no complex training required)" # because it is rule-based

# D) Stability: test with higher noise
V_noisier = V + np.random.normal(0, np.std(V)*0.3, V.shape)
edge_pred_noisy = detect_edges(V_noisier, threshold=5e-9)
acc_noisy = accuracy_score(y_true, edge_pred_noisy)

# E) Energy consumption (estimated for SNN)
# Assumption: each firing = 1 pJ (realistic value for nanoelectronic neurons)
num_spikes = np.sum(edge_pred)
energy_pJ = num_spikes * 1.0 # picoJoule

# ----------------------------
# 4. Display results
# ----------------------------
print("✅ Comprehensive test results of intelligent piezoelectric system")
print("="*50)
print(f"🎯 Accuracy: {acc:.4f} ({acc*100:.1f}%)")
print(f"🔍 Positive precision: {precision:.4f}")
print(f"📡 Recall: {recall:.4f}")
print(f"⚖️ F1-Score: {f1:.4f}")
print(f"⏱️ Latency: {latency_ms} ms")
print(f"🧠 Training: {trainable}")
print(f"🛡️ Stability (accuracy with 30% noise): {acc_noisy:.4f} ({acc_noisy*100:.1f}%)")
print(f"🔋 Energy consumption: {energy_pJ:.1f} pJ (≈ {num_spikes} firing)")

# Graph for better understanding
plt.figure(figsize=(12, 6))
plt.plot(time_sim, V_noisy, 'b', alpha=0.7, label='Piezo voltage (with noise)')
plt.scatter(time_sim[y_true==1], V_noisy[y_true==1], color='red', s=20, label='True spike')
plt.scatter(time_sim[edge_pred==1], V_noisy[edge_pred==1], color='green', s=15, marker='x', label='Predicted spike')
plt.title('Edge detection in signal Piezoelectric')
plt.xlabel('Time (s)')
plt.ylabel('Voltage (V)')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.savefig('/mnt/data/piezo_edge_detection.png')
plt.show()
import tensorflow as tf
from sklearn.metrics import accuracy_score, precision_recall_fscore_support
import time
import matplotlib.pyplot as plt

# ----------------------------
# 1. شبیه‌سازی داده‌های پیزوالکتریک (همان کد شما)
# ----------------------------
time_sim = np.linspace(0, 2, 1000)  # 2 ثانیه برای تست بهتر
sigma = 1e6 * np.abs(np.sin(2 * np.pi * 5 * time_sim))
d33 = 2.5e-10
t = 0.001
V = d33 * sigma * t  # ولتاژ خام

# ایجاد برچسب: 1 اگر sigma > 0.8e6 (ضربه)، 0 در غیر این صورت
y_true = (sigma > 0.8e6).astype(int)

# افزودن نویز واقع‌گرایانه (برای تست پایداری)
np.random.seed(42)
V_noisy = V + np.random.normal(0, np.std(V)*0.1, V.shape)

# ----------------------------
# 2. پیش‌پردازش: تشخیص لبه با گرادیان
# ----------------------------
def detect_edges(signal, threshold=1e-8):
    dv_dt = np.gradient(signal, time_sim)
    return (np.abs(dv_dt) > threshold).astype(int)

# آستانه بهینه‌شده برای داده‌های ما
edge_pred = detect_edges(V_noisy, threshold=5e-9)

# ----------------------------
# 3. معیارهای ارزیابی
# ----------------------------
# الف) دقت
acc = accuracy_score(y_true, edge_pred)
precision, recall, f1, _ = precision_recall_fscore_support(y_true, edge_pred, average='binary')

# ب) سرعت (تاخیر تشخیص)
# فرض: سیستم هر 10ms نمونه‌برداری می‌کند → تاخیر = 10ms
latency_ms = 10  # میلی‌ثانیه

# ج) آموزش (در اینجا ساده است، اما اگر از SNN با یادگیری بود، نرخ همگرایی را اندازه می‌گرفتیم)
trainable = "بله (بدون نیاز به آموزش پیچیده)"  # چون قاعده‌محور است

# د) پایداری: تست با نویز بالاتر
V_noisier = V + np.random.normal(0, np.std(V)*0.3, V.shape)
edge_pred_noisy = detect_edges(V_noisier, threshold=5e-9)
acc_noisy = accuracy_score(y_true, edge_pred_noisy)

# ه) مصرف انرژی (تخمینی برای SNN)
# فرض: هر شلیک = 1 pJ (مقدار واقع‌گرایانه برای نورون‌های نانو الکترونیک)
num_spikes = np.sum(edge_pred)
energy_pJ = num_spikes * 1.0  # picoJoule

# ----------------------------
# 4. نمایش نتایج
# ----------------------------
print("✅ نتایج تست جامع سیستم پیزوالکتریک هوشمند")
print("="*50)
print(f"🎯 دقت (Accuracy):          {acc:.4f} ({acc*100:.1f}%)")
print(f"🔍 دقت مثبت (Precision):   {precision:.4f}")
print(f"📡 فراخوانی (Recall):       {recall:.4f}")
print(f"⚖️  F1-Score:               {f1:.4f}")
print(f"⏱️  سرعت (تاخیر):          {latency_ms} ms")
print(f"🧠 آموزش:                  {trainable}")
print(f"🛡️  پایداری (دقت با نویز ۳۰٪): {acc_noisy:.4f} ({acc_noisy*100:.1f}%)")
print(f"🔋 مصرف انرژی:             {energy_pJ:.1f} pJ (≈ {num_spikes} شلیک)")

# نمودار برای درک بهتر
plt.figure(figsize=(12, 6))
plt.plot(time_sim, V_noisy, 'b', alpha=0.7, label='ولتاژ پیزو (با نویز)')
plt.scatter(time_sim[y_true==1], V_noisy[y_true==1], color='red', s=20, label='ضربه واقعی')
plt.scatter(time_sim[edge_pred==1], V_noisy[edge_pred==1], color='green', s=15, marker='x', label='ضربه تشخیص‌داده‌شده')
plt.title('تشخیص لبه در سیگنال پیزوالکتریک')
plt.xlabel('زمان (s)')
plt.ylabel('ولتاژ (V)')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.savefig('/mnt/data/piezo_edge_detection.png')
plt.show()[](url)

### CODE REVIEW

**Comment by NaderMalki** [comment]
#PCT/IR2025/050026 .Nader.maleki.al@gmail.com
